scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
+                return self.__pids
 
-        This will either use the commandline matcher or the target pid to find the process.
-        If no process is matched, None is returned.
+        matching_pids = self.get_pids_from_ps()
+        if self.__include_child_processes:
+            for matching_pid in matching_pids:
+                matching_pids.extend(
+                    self.get_pids_from_ps(subprocesses=True, ppid=matching_pid)
+                )
+        matching_pids = list(set(matching_pids))
+        return matching_pids
 
-        @return: The process id of the matching process, or None
-        @rtype: int or None
+    def should_poll_for_processes(self):
         """
-        sub_proc = None
-        if self.__commandline_matcher is not None:
-            try:
-                # Spawn a process to run ps and match on the command line.  We only output two
-                # fields from ps.. the pid and command.
-                sub_proc = Popen(['ps', 'ax', '-o', 'pid,command'],
-                                 shell=False, stdout=PIPE)
-
-                sub_proc.stdout.readline()
-                matching_pids = []
-                for line in sub_proc.stdout:
-                    line = line.strip()
-                    if line.find(' ') > 0:
-                        pid = int(line[:line.find(' ')])
-                        line = line[(line.find(' ') + 1):]
-                        if re.search(self.__commandline_matcher, line) is not None:
-                            matching_pids.append( pid )
+        We should poll for new processes in the following case:
 
-                total_pids = len( matching_pids )
-                if total_pids == 0:
-                    return None
-                elif total_pids == 1:
-                    return matching_pids[0]
+        * We do not currently have a process that we matched.
+        * This is a single process case, and the old process died.
+        * This is a multi-process case, and at least one process is alive and a certain time has expired.
+        :return: True if there is a need to poll, False otherwise
+        """
 
-                # if we are here then multiple command lines matched.  First check to
-                # see if self.__pid is in the list and if so return that.  Otherwise
-                # return the first element in the list
-                if self.__pid in matching_pids:
-                    self._logger.warning( "Multiple processes match the command '%s'.  Returning existing pid." % self.__commandline_matcher, limit_once_per_x_secs=300, limit_key='linux-process-monitor-existing-pid' )
-                    return self.__pid
+        if not self.__pids:
+            # no processes are currently being monitored, poll again
+            return True
+
+        if not self.__aggregate_multiple_processes:
+            # single process case
+            active_matched_pids = self.get_active_matched_pids(check_poll=False)
```

Comment:
You have `get_active_matched_pids` invoking this method, and this method invoking `get_active_matched_pids`.  That seems like a recipe for infinite recursion.

You should just simplify this and use `self.__pids` directly.